src/main/java/com/conveyal/taui/controllers/OpportunityDatasetsController.java

```diff
@@ -297,6 +267,57 @@ public static OpportunityDatasetUploadStatus createOpportunityDataset(Request re
         return grids;
     }
 
+    /**
+     * Respond to a request with a redirect to a downloadable file.
+     *
+     * @req should specify projectId, gridId, and an available download format (.tiff or .grid)
+     *
+     */
+    private static Object downloadOpportunityDataset (Request req, Response res) throws IOException {
+        // TODO check project membership
+        String projectId = req.params("projectId");
+        String gridId = req.params("gridId");
+        String gridPath = String.format("%s/%s", projectId, gridId);
+        String format = req.params("format");
+        String redirectText = req.queryParams("redirect");
+        boolean redirect;
+        redirect = redirectText == null || "".equals(redirectText) || parseBoolean(redirectText);
+
+        Grid grid;
+
+        if (!s3.doesObjectExist(BUCKET, String.format("%s.%s", gridPath, format))) {
+            // if this grid is not on S3 in the requested format, try to get the .grid format
+            if (!s3.doesObjectExist(BUCKET, String.format("%s.grid", gridPath))) {
+                throw new IllegalArgumentException("This grid does not exist.");
+            } else {
+                // get the grid and convert it to the requested format
+
+                S3Object s3Grid = s3.getObject(BUCKET, String.format("%s.grid", gridPath));
+                InputStream rawInput = s3Grid.getObjectContent();
```

Comment:
I believe you can use `grid = Grid.read` on this inputStream so you don't need to manually `readInt`s below. Unless there is a specific reason why you are doing it manually 🤔

Reply:
In the manual implementation here, north comes before west.

But in Grid.read, west comes before north:

![image](https://user-images.githubusercontent.com/2173529/33385878-b3bb879e-d4f7-11e7-8226-8d001315c715.png)

Reply:
oh boy...that's not good

Reply:
Client will expect west before north: https://github.com/conveyal/browsochrones/blob/dev/lib/grid.js

Reply:
So shouldn't we just fix this in an additional commit - standardize the ordering of fields within the file format?

Reply:
Yeah, my manual implementation was just wrong 🙄 

I'll take it as a lesson to look for and use the existing methods.